OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,274 @@
+    }
+    
+    /** Return true if the DataTable contains a column for the label. */
+    bool hasColumn(const std::string& label) const override{
+        return (_columnLabels.findIndex(label) >= 0);
+    }
+
+    size_t getColumnIndex(const std::string& label) const override {
+        return _columnLabels.findIndex(label);
+    }
+
+    /** Read-only access to a row of data in the table as a RowVector.*/
+    const SimTK::RowVectorView_<DataType> getRow(size_t row) const {
+        return _data[int(row)];
+    }
+
+    /** Write access to a column of data as a RowVector */
+    SimTK::RowVectorView_<DataType> updRow(size_t row) {
+        return _data[int(row)];
+    }
+
+    /** Read-only access to a column of data in the table as a Vector.*/
+    const SimTK::VectorView_<DataType> getColumn(size_t col) const {
+        return _data(int(col));
+    }
+    /** Write access to a column of data as a Vector */
+    SimTK::VectorView_<DataType> updColumn(size_t col) {
+        return _data(int(col));
+    }
+
+    /** Read-only access to an element of DataType in the table by row and
+        column indices.*/
+    const DataType& getElement(size_t row, size_t col) const {
+        return _data(int(row), int(col));
+    }
+
+    /** Read-only access to the table as a Matrix */
+    const SimTK::Matrix_<DataType>& getAsMatrix() const {
+        return _data;
+    }
+
+    /** Append a row of data as a RowVector to the table. */
+    void appendRow(const SimTK::RowVector_<DataType>& row) {
+        // ncols specified by the table unless it is zero, in which
+        // case allow the first row appended to dictate its size
+        int ncols = _data.ncol() == 0 ? row.size() : _data.ncol();
+        SimTK_ASSERT_ALWAYS(row.size() == ncols, "DataTable::appendRow() "
+            "row length does match number of columns.");
+        SimTK_ASSERT_ALWAYS(row.size() > 0, "DataTable::appendRow() "
+            "row is empty.");
+        _data.resizeKeep(_data.nrow() + 1, ncols);
+        _data[_data.nrow() - 1].updAsRowVector() = row;
+    }
+
+    /** Append another data table's rows to this table. If the number of
+        columns are incompatible it will throw and exception. */
+    void appendDataTable(const DataTable_& table) {
+        size_t nrows = table.getNumRows();
+        if (getNumCols() != table.getNumCols()) {
+            throw Exception("DataTable::appendDataTable() cannot append a "
+                " DataTable with a different number of columns.");
+        }
+        // resize once to tack on the new table
+        size_t offset = _data.nrow();
+        _data.resizeKeep(int( offset + nrows), int(getNumCols()) );
+        for (size_t i = 0; i < nrows; ++i){
+            _data[int(i + offset)].updAsRowVector() = table.getRow(i);
+        }
+    }
+
+    /** Dump the contents of table and strings to an output stream.
+        Primarily used for debugging purposes.*/
+    void dumpToStream(std::ostream &out) const override {
```

Comment:
I am not sure if we intend to support de-serialization of the DataTable. Are we ?
If yes, then it would probably be better to rely on third party serialization libraries to do this. Example -- Google protobuf library.
I would guess that a user would want to read the data into another software packages like Matlab, R, SPSS, SAS etc. In this case, serializing to a CSV would be very helpful.